tests/jballerina-semtype-port-test/src/test/java/io/ballerina/semtype/port/test/SemTypeResolver.java

```diff
@@ -0,0 +1,383 @@
+        if (moduleLevelDef == null) {
+            throw new IllegalStateException("unknown type: " + name);
+        }
+
+        if (moduleLevelDef.getKind() == NodeKind.TYPE_DEFINITION) {
+            return resolveTypeDefn(cx, mod, (BLangTypeDefinition) moduleLevelDef, depth);
+        } else if (moduleLevelDef.getKind() == NodeKind.CONSTANT) {
+            BLangConstant constant = (BLangConstant) moduleLevelDef;
+            return resolveTypeDefn(cx, mod, constant.associatedTypeDefinition, depth);
+        } else {
+            throw new UnsupportedOperationException("constants and class defns not implemented");
+        }
+    }
+
+    private SemType resolveIntSubtype(String name) {
+        // TODO: support MAX_VALUE
+        return switch (name) {
+            case "Signed8" -> SemTypes.SINT8;
+            case "Signed16" -> SemTypes.SINT16;
+            case "Signed32" -> SemTypes.SINT32;
+            case "Unsigned8" -> SemTypes.UINT8;
+            case "Unsigned16" -> SemTypes.UINT16;
+            case "Unsigned32" -> SemTypes.UINT32;
+            default -> throw new UnsupportedOperationException("Unknown int subtype: " + name);
+        };
+    }
+
+    private SemType resolveXmlSubtype(String name) {
+        return switch (name) {
+            case "Element" -> SemTypes.XML_ELEMENT;
+            case "Comment" -> SemTypes.XML_COMMENT;
+            case "Text" -> SemTypes.XML_TEXT;
+            case "ProcessingInstruction" -> SemTypes.XML_PI;
+            default -> throw new IllegalStateException("Unknown XML subtype: " + name);
+        };
+    }
+
+    private SemType resolveSingletonType(BLangFiniteTypeNode td) {
+        return resolveSingletonType(td.valueSpace);
+    }
+
+    private SemType resolveSingletonType(List<BLangExpression> valueSpace) {
+        List<SemType> types = new ArrayList<>();
+        for (BLangExpression bLangExpression : valueSpace) {
+            types.add(resolveSingletonType((BLangLiteral) bLangExpression));
+        }
+
+        Iterator<SemType> iter = types.iterator();
+        SemType u = iter.next();
+        while (iter.hasNext()) {
+            u = SemTypes.union(u, iter.next());
+        }
+        return u;
+    }
+
+    private SemType resolveSingletonType(BLangLiteral literal) {
+        return resolveSingletonType(literal.value, literal.getDeterminedType().getKind());
+    }
+
+    private SemType resolveSingletonType(Object value, TypeKind targetTypeKind) {
+        return switch (targetTypeKind) {
+            case NIL -> PredefinedType.NIL;
+            case BOOLEAN -> SemTypes.booleanConst((Boolean) value);
+            case INT, BYTE -> SemTypes.intConst(((Number) value).longValue());
+            case FLOAT -> {
+                double doubleVal;
+                if (value instanceof Long) {
+                    doubleVal = ((Long) value).doubleValue();
+                } else if (value instanceof Double) {
+                    doubleVal = (double) value;
```

Comment:
I think we can use patter matching here to avoid explicit casts
```suggestion
                if (value instanceof Long longValue) {
                    doubleVal = longValue.doubleValue();
                } else if (value instanceof Double doubleValue) {
                    doubleVal = doubleValue;
```